### PR TITLE
initial commit, will add tests in next commit

### PR DESCRIPTION
**Pull request by SatishThareja**
Adding test impact data collector in testsettings and run settings for node implementation.

### CODE REVIEW

File: Tasks/VsTest/vstest.ts [hunk 1]
```diff
@@ -105,83 +105,90 @@ function invokeVSTest(testResultsDirectory: string): Q.Promise<number> {
...
-
-                            testAssemblyFiles.forEach(function(testAssembly) {
-                                var testAssemblyPath = testAssembly;
-                                //To maintain parity with the behaviour when test assembly was filepath, try to expand it relative to build sources directory.
-                                if (sourcesDirectory && !pathExistsAsFile(testAssembly)) {
-                                    var expandedPath = path.join(sourcesDirectory, testAssembly);
-                                    if (pathExistsAsFile(expandedPath)) {
-                                        testAssemblyPath = expandedPath;
-                                    }
+                    setupRunSettingsFileForTestImpact(vsVersion,overriddenSettingsFile)
```
Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: if for some reason this fails we shuld continue to run all tests without TIA on. this shuld be a blocker

File: Tasks/VsTest/vstest.ts [hunk 2]
```diff
@@ -343,8 +350,249 @@ function getTestResultsDirectory(settingsFile: string, defaultResultsDirectory:
...
+             assemblyQualifiedName: getTIAssemblyQualifiedName(vsVersion),
+             friendlyName: getTIFriendlyName() };
+}
+
+function isTestImapctCollectorPresent(dataCollectorArray) {
+    var found = false;
+    var tiaFriendlyName = getTIFriendlyName();
+    for (var i = 0; i < dataCollectorArray.length; i++) {
+        try {
+            if (dataCollectorArray[i].$.friendlyName === tiaFriendlyName) {
```
Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: will this be ignore case check??

File: Tasks/VsTest/vstest.ts [hunk 3]
```diff
@@ -343,8 +350,249 @@ function getTestResultsDirectory(settingsFile: string, defaultResultsDirectory:
...
+                if (err) {
+                    tl.warning(tl.loc('ErrorWhileReadingRunSettings', err));
+                    tl.debug(exitErrorMessage);
+                    defer.resolve(settingsFile);
+                    return defer.promise;
+                }
+                if (result.RunSettings === undefined) {
+                    tl.warning(tl.loc('FailedToSetTestImpactCollectorRunSettings'));
+                    defer.resolve(settingsFile);
+                    return defer.promise;
```
Comment (acesiddhu) [inline Tasks/VsTest/vstest.ts]: this shuld just be a warning and we shuld continue after this and not fail the task

Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: this error prints a warning on the console. the execution continues with the settings file unchanged.

File: Tasks/VsTest/vstest.ts (hunk 1 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: The implementation is same for test impact settings as it for run in parallel option. Tests will continue with original test settings file.

File: Tasks/VsTest/vstest.ts (hunk 2 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: Converting to upper case before comparison.

File: Tasks/VsTest/vstest.ts [hunk 4]
```diff
@@ -343,8 +352,218 @@ function getTestResultsDirectory(settingsFile: string, defaultResultsDirectory:
...
+
+function GetTIAssemblyQualifiedName(vsVersion) {
+    return "Microsoft.VisualStudio.TraceCollector.TestImpactDataCollector, Microsoft.VisualStudio.TraceCollector, Version=" + vsVersion + ".0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a";
+}
+
+function GetTIFriendlyName() {
+    return "Test Impact";
+}
+
+function GetTestImpactAttributes(vsVersion) {
```
Comment (AshwiniChalla) [inline Tasks/VsTest/vstest.ts]: vsVersion: number declare types everywhere

File: Tasks/VsTest/vstest.ts [hunk 5]
```diff
@@ -343,8 +352,218 @@ function getTestResultsDirectory(settingsFile: string, defaultResultsDirectory:
...
+                        if (!dataCollectorArray) {
+                            tl.debug("Updating testsettings file from DataCollector node");
+                            result.TestSettings.Execution[0].AgentRule[0].DataCollectors[0] = { DataCollector: {} };
+                            dataCollectorNode = result.TestSettings.Execution[0].AgentRule[0].DataCollectors[0].DataCollector;
+                        }
+                        else {
+                            if (!IsTestImapctCollectorPresent(dataCollectorArray)) {
+                                tl.debug("Updating testsettings file, adding a DataCollector node");
+                                dataCollectorArray.push({});
+                                result.TestSettings.Execution[0].AgentRule[0].DataCollectors[0].DataCollector = dataCollectorArray;
```
Comment (AshwiniChalla) [inline Tasks/VsTest/vstest.ts]: Can we be sure that this wont throw null ref exception ?

File: Tasks/VsTest/vstest.ts [hunk 6]
```diff
@@ -343,8 +352,218 @@ function getTestResultsDirectory(settingsFile: string, defaultResultsDirectory:
...
+                                dataCollectorNode = dataCollectorArray[dataCollectorArray.length - 1];
```
Comment (AshwiniChalla) [inline Tasks/VsTest/vstest.ts]: nit: dataCollectorArray.last

File: Tasks/VsTest/vstest.ts [hunk 7]
```diff
@@ -343,8 +352,218 @@ function getTestResultsDirectory(settingsFile: string, defaultResultsDirectory:
...
+                });
+            })
+                .fail(function (err) {
+                tl.warning(err);
+                tl.debug(exitErrorMessage);
+                defer.resolve(settingsFile);
+            });
+        }
+        else if (!settingsFile || settingsFile.split('.').pop().toLowerCase() != "runsettings" || !pathExistsAsFile(settingsFile)) {
+            tl.debug("No settings file provided or the provided settings file does not exist. Creating test settings file for enabling test impact data collector.");
```
Comment (AshwiniChalla) [inline Tasks/VsTest/vstest.ts]: typo: creating run settings

File: Tasks/VsTest/vstest.ts [hunk 8]
```diff
@@ -456,22 +675,25 @@ function resetRunInParallel() {
...
-
     var regPath = "HKLM\\SOFTWARE\\Microsoft\\VisualStudio";
     regedit.list(regPath).on('data', function(entry) {
         if (entry && entry.data && entry.data.keys) {
             var subkeys = entry.data.keys;
             var versions = getFloatsFromStringArray(subkeys);
             if (versions && versions.length > 0) {
-                versions.sort();
-                defer.resolve(parseFloat(versions[versions.length - 1]));
+                versions.sort(compareNumbers);
```
Comment (AshwiniChalla) [inline Tasks/VsTest/vstest.ts]: compareNumbers is optional

**Pull request closed without merging**